benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu

```diff
@@ -0,0 +1,251 @@
+
+  auto num_streams = state.range(0);
+  auto num_kernels = state.range(1);
+  auto do_prewarm  = state.range(2);
+
+  auto stream_pool = rmm::cuda_stream_pool(num_streams);
+
+  if (do_prewarm) { run_prewarm(stream_pool, mr.get()); }
+
+  for (auto _ : state) {
+    run_test(num_kernels, stream_pool, mr.get());
+    cudaDeviceSynchronize();
+  }
+
+  state.SetItemsProcessed(state.iterations() * num_kernels);
+
+  rmm::mr::set_current_device_resource(nullptr);
+}
+
+inline auto make_cuda() { return std::make_shared<rmm::mr::cuda_memory_resource>(); }
+
+inline auto make_cuda_async() { return std::make_shared<rmm::mr::cuda_async_memory_resource>(); }
+
+inline auto make_pool()
+{
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(make_cuda());
+}
+
+inline auto make_arena()
+{
+  return rmm::mr::make_owning_wrapper<rmm::mr::arena_memory_resource>(make_cuda());
+}
+
+inline auto make_binning()
+{
+  auto pool = make_pool();
+  // Add a binning_memory_resource with fixed-size bins of sizes 256, 512, 1024, 2048 and 4096KiB
+  // Larger allocations will use the pool resource
+  auto mr = rmm::mr::make_owning_wrapper<rmm::mr::binning_memory_resource>(pool, 18, 22);
+  return mr;
+}
+
+static void benchmark_range(benchmark::internal::Benchmark* b)
+{
+  b  //
+    ->RangeMultiplier(2)
+    ->Ranges({{1, 4}, {4, 4}, {false, true}})
+    ->Unit(benchmark::kMicrosecond);
+}
+
+MRFactoryFunc get_mr_factory(std::string resource_name)
+{
+  if (resource_name == "cuda") { return &make_cuda; }
+#ifdef RMM_CUDA_MALLOC_ASYNC_SUPPORT
+  if (resource_name == "cuda_async") { return &make_cuda_async; }
+#endif
+  if (resource_name == "pool") { return &make_pool; }
+  if (resource_name == "arena") { return &make_arena; }
+  if (resource_name == "binning") { return &make_binning; }
+
+  std::cout << "Error: invalid memory_resource name: " << resource_name << std::endl;
+
+  RMM_FAIL();
+}
+
+void declare_benchmark(std::string name)
+{
+  if (name == "cuda") {
+    BENCHMARK_CAPTURE(BM_MultiStreamAllocations, cuda, &make_cuda)  //
+      ->Apply(benchmark_range);
+    return;
+  }
+
+#ifdef RMM_CUDA_MALLOC_ASYNC_SUPPORT
+  if (name == "cuda_async") {
+    BENCHMARK_CAPTURE(BM_MultiStreamAllocations, cuda_async, &make_cuda_async)  //
+      ->Apply(benchmark_range);
+    return;
+  }
+#endif
+
+  if (name == "pool") {
+    BENCHMARK_CAPTURE(BM_MultiStreamAllocations, pool_mr, &make_pool)  //
+      ->Apply(benchmark_range);
+    return;
+  }
+
+  if (name == "arena") {
+    BENCHMARK_CAPTURE(BM_MultiStreamAllocations, arena, &make_arena)  //
+      ->Apply(benchmark_range);
+    return;
+  }
+
+  if (name == "binning") {
+    BENCHMARK_CAPTURE(BM_MultiStreamAllocations, binning, &make_binning)  //
+      ->Apply(benchmark_range);
+    return;
+  }
+
+  std::cout << "Error: invalid memory_resource name: " << name << std::endl;
+}
+
+void run_profile(std::string resource_name, int kernel_count, int stream_count, bool prewarm)
```

Comment:
Is this needed? Can't you do the same thing via GBench command line args?

Reply:
Not sure what you mean by "this". gbench runs multiple times no matter what. We want a way to run only once for profiling.

Reply:
> We want a way to run only once for profiling.

Right, that's what I meant. I thought there was a "num_iterations` gbench option. 

Related https://github.com/NVIDIA/nvbench/issues/10

Reply:
I haven't found an option in gbench to limit the number of iterations. I looked briefly prior to implementing it this way, but gbench documentation is... not great?

Reply:
gbenchmark does not have a way to control number of iterations. Here's one of the authors explanations of why. https://stackoverflow.com/a/61888885/749748